File: algorithms/javascript/merge_sort.js
const mergeSort = (list) => {
    const len = list.length;
    if (len == 1) {
        return list
    };
    // get middle item 
    const mid = Math.ceil(len / 2);

    let left = list.slice(0, mid);
    let right = list.slice(mid, len);

    left = mergeSort(left);
    right = mergeSort(right);

    return merge(left, right);
}

const merge = (left, right) => {
    const sorted = [];
    while (left.length > 0 && right.length > 0) {
        const lItem = left[0];
        const rItem = right[0];
        if (lItem > rItem) {
            sorted.push(rItem);
            right.shift();
        } else {
            sorted.push(lItem);
            left.shift();
        }
    }
    // if arr's not same length, add the rest
    let pushRemaining = (arr) => {
        while (arr.length !== 0) {
            sorted.push(arr.shift());
        }
    };

    pushRemaining(right);
    pushRemaining(left);

    return sorted
}


let items = [3, 2, 4, 1, 5, 7, 8, 6];
console.log("before sort", items);
items = mergeSort(items);
console.log("after sort", items);

// Merge sort uses a divide-and-conquer approach to sort elements in
// an array. Basically, what this means is that instead of dealing with
// the array as a whole, it continually splits it in half until both
// halves are sorted, then the halves are merged into one solved list.

File: algorithms/python/divide_and_conquer/mergesort.py
def merge(arr, start, mid, end):
    left = arr[start:mid+1]
    right = arr[mid+1:end+1]
    i, j, k = 0, 0, start

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        arr[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        arr[k] = right[j]
        j += 1
        k += 1

    return arr

def mergesort(arr, l, r):
    """
    >>> mergesort([3,2,1],0,2)
    [1, 2, 3]
    >>> mergesort([3,2,1,0,1,2,3,5,4],0,8)
    [0, 1, 1, 2, 2, 3, 3, 4, 5]
    """

    if l < r:
        m = (l + r) // 2

        mergesort(arr, l, m)
        mergesort(arr, m+1, r)
        merge(arr, l, m, r)

        return arr

if __name__ == "__main__":
    arr = [3, 2, 1, 0, 1, 2, 3, 5, 4]
    print(arr)
    print(mergesort(arr, 0, len(arr)))
